enumerables.rb

```diff
@@ -0,0 +1,153 @@
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    incr = 0
+    arr_kind = self.class
+    if arr_kind == Hash
+      arr = flatten
+      incr = 2
+    else
+      arr = to_a
+      incr = 1
+    end
+
+    i = 0
+    while i < arr.length
+      arr_kind == Hash ? yield(arr[i], arr[i + 1]) : yield(arr[i])
+      i += incr
+    end
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    arr = to_a
+
+    for i in 0...arr.length
+      yield(arr[i], i)
+    end
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    kind = self.class
+    arr = to_a
+
+    if kind == Hash
+      match = {}
+      flat = arr.flatten
+      i = 0
+      while i < flat.length
+        found = yield(flat[i], flat[i + 1])
+        match[flat[i]] = flat[i + 1] if found
+        i += 2
+      end
+    else
+      match = []
+      arr.my_each do |val|
+        found = yield(val)
+        match << val if found
+      end
+    end
+    match
+  end
+
+  def my_all?(pattern = nil)
+    arr = to_a
+    all_true = true
+
+    arr.my_each do |v|
+      all_true = if pattern
+                   !!(v =~ pattern)
```

Comment:
Reconsider this line

Reply:
I removed the `!!` and used `#match` instead of `=~`. It also checks for classes as well. I hope this is enough :pray: